KAFKA-7778: Add KTable.suppress to Scala API

ouertani opened the pull request:
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*
Add KTable.suppress to Scala API
https://issues.apache.org/jira/browse/KAFKA-7778

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [X] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [X ] Verify documentation (including upgrade notes)

ouertani committed:
add suppress

ouertani committed:
add suppress

guozhangwang commented:
Jenkins failure seems related to `spotless`. Could you run unit test locally and see if it passed? Also could you add the corresponding unit test?

guozhangwang commented:
cc @vvcephei

guozhangwang committed:
MINOR: Bump js template to 2.2 (#6086)

Bump js template to 2.2; update gradle.properties

Reviewers: Jason Gustafson <jason@confluent.io>

bbejeck committed:
MINOR: Remove sleep calls and ignore annotation from streams upgrade test (#6046)

The StreamsUpgradeTest::test_upgrade_downgrade_brokers used sleep calls in the test which led to flaky test performance and as a result, we placed an @ignore annotation on the test. This PR uses log events instead of the sleep calls hence we can now remove the @ignore setting.

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

hackerwin7 committed:
KAFKA-7755; Look up client host name since DNS entry may have changed (#6049)

Lookup client host name after every full iteration through the addresses returned.

Reviewers: Loïc Monney <loicmonney@github.com>, Edoardo Comar <ecomar@uk.ibm.com>, Rajini Sivaram <rajinisivaram@googlemail.com>

vvcephei commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala:
```diff
@@ -159,6 +159,16 @@ class KTable[K, V](val inner: KTableJ[K, V]) {
     inner.toStream[KR](mapper.asKeyValueMapper)
 
+  /**
+    * Suppress some updates from this changelog stream, determined by the supplied {@link Suppressed} configuration.
```
`@link` is Javadoc, not valid Scaladoc. In scaladoc, it's `[[Suppressed]]`:

```suggestion
    * Suppress some updates from this changelog stream, determined by the supplied [[Suppressed]] configuration.
```

vvcephei commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala:
```diff
@@ -159,6 +159,16 @@ class KTable[K, V](val inner: KTableJ[K, V]) {
+    * This controls what updates downstream table and stream operations will receive.
+    *
+    * @param suppressed Configuration object determining what, if any, updates to suppress
+    * @return A new KTable with the desired suppression characteristics.
```
```suggestion
    * @return A new [[KTable]] with the desired suppression characteristics.
```

vvcephei reviewed (changes requested):
Hi @ouertani ,

Thanks for picking this up!

I commented on a couple of scaladoc syntax things.

The tests failed because the method header is missing some spaces. You can look at the console output of the build to see what it should be, or you can just run `./gradlew spottlessApply`, and it'll reformat it for you.

Experience has taught us that we really need unit tests for the scala API as well, even thought your implementation is "obviously correct". For example, in the past, we've seen problems with the type bounds.

I think a simple test that verifies:
* `untilWindowCloses` with a `KTable[Windowed[_],_]`
* `untilTimeLimit` with a `KTable[Windowed[_],_]`
* `untilTimeLimit` with a `KTable[(not Windowed) ,_]`

And, unfortunately, we can't test code that doesn't compile, but please confirm that the following fails the type check:
* `untilWindowCloses` with a `KTable[(not Windowed),_]`

ryannedolan commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala:
```diff
@@ -159,6 +159,16 @@ class KTable[K, V](val inner: KTableJ[K, V]) {
+    * @param suppressed Configuration object determining what, if any, updates to suppress
+    * @return A new KTable with the desired suppression characteristics.
+    */
+  def suppress(suppressed: Suppressed[_ >: K]):KTable[K,V] = inner.suppress(suppressed)
```
```suggestion
  def suppress(suppressed: Suppressed[_ >: K]): KTable[K, V] = inner.suppress(suppressed)
```

ouertani committed:
KAFKA-7768: Add version to java html urls (#6094)

Reviewers: Guozhang Wang <wangguoz@gmail.com>

cmccabe committed:
KAFKA-7051: Improve the efficiency of ReplicaManager (#5206)

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Dong Lin <lindong28@gmail.com>

cmccabe committed:
KAFKA-7051: Improve the efficiency of ReplicaManager (fixup)

vvcephei committed:
KAFKA-7768: Use absolute paths for javadoc (#6100)

Reviewers: Guozhang Wang <wangguoz@gmail.com>

hachikuji committed:
KAFKA-7773; Add end to end system test relying on verifiable consumer (#6070)

This commit creates an EndToEndTest base class which relies on the verifiable consumer. This will ultimately replace ProduceConsumeValidateTest which depends on the console consumer. The advantage is that the verifiable consumer exposes more information to use for validation. It also allows for a nicer shutdown pattern. Rather than relying on the console consumer idle timeout, which requires a minimum wait time, we can halt consumption after we have reached the last acked offsets. This should be more reliable and faster. The downside is that the verifiable consumer only works with the new consumer, so we cannot yet convert the upgrade tests. This commit converts only the replication tests and a flaky security test to use EndToEndTest.

chia7712 committed:
KAFKA-7253; The returned connector type is always null when creating connector (#5470)

The null map returned from the current snapshot causes the null type in response. The connector class name can be taken from the config of request instead since we require the config should contain the connector class name.

Reviewers: Jason Gustafson <jason@confluent.io>

mjsax committed:
MINOR: code cleanup (#6056)

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

bbejeck committed:
MINOR: Put states in proper order, increase timeout for starting (#6105)

Reviewers: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

bob-barrett committed:
KAFKA-6833; Producer should await metadata for unknown partitions (#6073)

This patch changes the behavior of KafkaProducer.waitOnMetadata to wait up to max.block.ms when the partition specified in the produce request is out of the range of partitions present in the metadata. This improves the user experience in the case when partitions are added to a topic and a client attempts to produce to one of the new partitions before the metadata has propagated to the brokers. Tested with unit tests.

Reviewers: Arjun Satish <arjun@confluent.io>, Jason Gustafson <jason@confluent.io>

hachikuji committed:
KAFKA-7799; Fix flaky test RestServerTest.testCORSEnabled (#6106)

The test always fails if testOptionsDoesNotIncludeWadlOutput is executed before testCORSEnabled. It seems the problem is the use of the system property. Perhaps there is some static caching somewhere.

Reviewers: Randall Hauch <rhauch@gmail.com>, Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
MINOR: code cleanup (#6055)

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

mjsax committed:
MINOR: code cleanup (#6057)

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

mjsax committed:
MINOR: code cleanup (#6053)

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Ryanne Dolan <ryannedolan@gmail.com>, Guozhang Wang <guozhang@confluent.io>

stanislavkozlovski committed:
MINOR: Log successful/failed authentications with socket information (#5856)

Use `info` for failed authentications and `debug` for successful ones.

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>, Ismael Juma <ismael@juma.me.uk>

apovzner committed:
KAFKA-7786; Ignore OffsetsForLeaderEpoch response if epoch changed while request in flight (#6101)

There is a race condition in ReplicaFetcherThread, where we can update PartitionFetchState with the new leader epoch (same leader) before handling the OffsetsForLeaderEpoch response with FENCED_LEADER_EPOCH error which causes removing partition from partitionStates, which in turn causes no fetching until the next LeaderAndIsr. 

This patch adds logic to ensure that the leader epoch doesn't change while an OffsetsForLeaderEpoch request is in flight (which could happen with back-to-back leader elections). If it has changed, we ignore the response.

Also added toString() implementation to PartitionData, because some log messages did not show useful info which I found while investigating the above system test failure.

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jason Gustafson <jason@confluent.io>

bbejeck committed:
MINOR: Add 2.1 version metadata upgrade (#6111)

Updated the test_metadata_upgrade test. To enable using the 2.1 version I needed to add config change to the StreamsUpgradeTestJobRunnerService to ensure the ductape passes proper args when starting the StreamsUpgradeTest

For testing, I ran the test_metadata_upgrade test and all versions now pass http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-01-09--001.1547049873--bbejeck--MINOR_add_2_1_version_metadata_upgrade--a450c68/report.html

Reviewers: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

ijuma committed:
MINOR: Update dependencies for Kafka 2.2 (#6116)

- Scala 2.12.7 -> 2.12.8
- Gradle 5.0 -> 5.1
- Jetty 9.4.12 -> 9.4.14
- Rat 0.12 -> 0.13
- Patch bumps for easymock, jackson, powermock
- Patch bumps for gradle plugins: shadow, spotbugs, dependency-check, spotless

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

omkreddy committed:
KAFKA-5994; Log ClusterAuthorizationException for all ClusterAction requests

Author: Manikumar Reddy <manikumar.reddy@gmail.com>

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>

Closes #5021 from omkreddy/KAFKA-5994-CLUSTER-AUTH

bbejeck committed:
MINOR: Remove throwing exception if not found from describe topics (#6112)

We recently improved the handling of the InternalTopicManager retries with #6085. The AdminClient will throw an InvalidTopicException if the topic is not found. We need to ignore that exception as when calling AdminClient#describe we may not have had a chance to create the topic yet, especially with the case of internal topics

I've created a new test asserting that when an InvalidTopicException is thrown when the topic is not found we continue on.

Reviewers: John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

vvcephei committed:
KAFKA-7741: Streams exclude javax dependency (#6121)

As documented in https://issues.apache.org/jira/browse/KAFKA-7741,
the javax dependency we receive transitively from connect is incompatible
with SBT builds.

Streams doesn't use the portion of Connect that needs the dependency,
so we can fix the builds by simply excluding it.

Reviewers: Bill Bejeck <bill@confluent.io>, Guozhang Wang <guozhang@confluent.io>

likan999 committed:
KAFKA-6627: Prevent config default values overriding ones specified through --producer-property on command line. (#6084)

* KAFKA-6627: Prevent config default values overriding ones specified through --producer-property on command line.

In Console{Producer,Consumer}, extraProducerProps (options specified in
--producer-property) is applied first, then overriden unconditionally,
even if the value is not specified explicitly (and default value is
used). This patch fixes it so that it doesn't override the existing
value set by --producer-property if it is not explicitly specified.

The contribution is my original work and I license the work to the
project under the project's open source license.

Reviewers: Sriharsha Chintalapani <sriharsha@apache.org>

bbejeck committed:
MINOR: Need to have same wait as verify timeout broker upgrade downgrade (#6127)

When I originally refactored the streams_upgrade_test#upgrade_downgrade_brokers test I removed the wait call which would wait for up 24 minutes for the SmokeTestDriver class to publish and verify all of its records.

Since most of the tests run in two minutes or less I set the monitor_log timeout to three minutes. However, the SmokeTestDriver#verify method allows up to six minutes to consume all records before verifying the monitor_log timeout needs to be greater than 6 minutes. I've set the timeout to 8 minutes.

Additionally, the steps needed to update the streams_upgrade_test should be documented as there are several components that need to get updated. So I've documented those steps here on the test as a giant comment.

Reviewers: Guozhang Wang <wangguoz@gmail.com>

stanislavkozlovski committed:
KAFKA-7790: Fix Bugs in Trogdor Task Expiration (#6103)

The Trogdor Coordinator now overwrites a task's startMs to the time it received it if startMs is in the past.

The Trogdor Agent now correctly expires a task after the expiry time (startMs + durationMs) passes. Previously, it would ignore startMs and expire after durationMs milliseconds of local start of the task.

Reviewed-by: Colin P. McCabe <cmccabe@apache.org>

hachikuji committed:
HOTFIX: Compilation error in CommandLineUtils (#6131)

This was broken by #6084. The syntax works with Scala 2.12, but not 2.11.

Reviewers: Colin Patrick McCabe <colin@cmccabe.xyz>

dongjinleekr committed:
KAFKA-7808: AdminClient#describeTopics should not throw InvalidTopic if topic name is not found (#6124)

* Update KafkaAdminClient#describeTopics to throw UnknownTopicOrPartitionException.

* Remove unused method: WorkerUtils#getMatchingTopicPartitions.

* Add some JavaDoc.

Reviewed-by: Colin P. McCabe <cmccabe@apache.org>, Ryanne Dolan <ryannedolan@gmail.com>

Commit:
MINOR: Support choosing different JVMs when running integration tests

+ Add a parameter to the ducktap-ak to control the OpenJDK base image.
+ Fix a few issues of using OpenJDK:11 as the base image.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

Author: Xi Yang <xi@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #6071 from yangxi/ducktape-jdk

cmccabe committed:
KAFKA-7609; Add Protocol Generator for Kafka (#5893)

This patch adds a framework to automatically generate the request/response classes for Kafka's protocol. The code will be updated to use the generated classes in follow-up patches. Below is a brief summary of the included components:

**buildSrc/src**
The message generator code is here.  This code is automatically re-run by gradle when one of the schema files changes.  The entire directory is processed at once to minimize the number of times we have to start a new JVM.  We use Jackson to translate the JSON files into Java objects.

**clients/src/main/java/org/apache/kafka/common/protocol/Message.java**
This is the interface implemented by all automatically generated messages.

**clients/src/main/java/org/apache/kafka/common/protocol/MessageUtil.java**
Some utility functions used by the generated message code.

**clients/src/main/java/org/apache/kafka/common/protocol/Readable.java, Writable.java, ByteBufferAccessor.java**
The generated message code uses these classes for writing to a buffer.

**clients/src/main/message/README.md**
This README file explains how the JSON schemas work.

**clients/src/main/message/\*.json**
The JSON files in this directory implement every supported version of every Kafka API.  The unit tests automatically validate that the generated schemas match the hand-written schemas in our code.  Additionally, there are some things like request and response headers that have schemas here.

**clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java**
I added an optimization here for empty sets.  This is useful here because I want all messages to start with empty sets by default prior to being loaded with data.  This is similar to the "empty list" optimizations in the `java.util.ArrayList` class.

Reviewers: Stanislav Kozlovski <stanislav_kozlovski@outlook.com>, Ismael Juma <ismael@juma.me.uk>, Bob Barrett <bob.barrett@outlook.com>, Jason Gustafson <jason@confluent.io>

kamalcph committed:
KAFKA-7781; Add validation check for retention.ms topic property.

Using AdminClient#alterConfigs, topic `retention.ms` property can be assigned to a value lesser than -1. This leads to inconsistency while describing the topic configuration. We should not allow values lesser than -1.

Author: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>,Matthias J. Sax <matthias@confluent.io>

Closes #6082 from kamalcph/KAFKA-7781

MayureshGharat committed:
KAFKA-4453 : Added code to separate controller connections and requests from the data plane (#5921)

KIP-291 Implementation : Added code to separate controller connections and requests from the data plane.

Tested with local deployment that the controller request are handled by the control plane and other requests are handled by the data plane.
Also added unit tests in order to test the functionality.

Author: Lucas Wang <luwang@linkedin.com>, 
Author: Mayuresh Gharat <gharatmayuresh15@gmail.com>

Reviewers: Joel Koshy <jjkoshy@gmail.com>, Jun Rao <junrao@gmail.com>

vvcephei committed:
MINOR: clarify the record selection algorithm and stream-time definition (#6128)

The existing javadoc for PartitionGroup is a little confusing.
It's relatively important for these concepts to be clear, since
they form the basis for stream-time in Kafka Streams.

Reviewers: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

bbejeck committed:
MINOR: Put state args in correct order named repartition test (#6114)

Another system test that needs to be updated with states in the correct order

Reviewers: Guozhang Wang <wangguoz@gmail.com>

ijuma committed:
HOTFIX: Fix Properties.putAll compiler error when compiling with Java 11 (#6140)

asasvari committed:
MINOR: Fix typo in security.html (#6141)

kafka-delegation-tokens.sh is the name of the script

mjsax committed:
MINOR: code cleanup (#6054)

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>, Ryanne Dolan <ryannedolan@gmail.com>, Ismael Juma <ismael@confuent.io>

wicknicks committed:
KAFKA-7503: Connect integration test harness

Expose a programmatic way to bring up a Kafka and Zk cluster through Java API to facilitate integration tests for framework level changes in Kafka Connect. The Kafka classes would be similar to KafkaEmbedded in streams. The new classes would reuse the kafka.server.KafkaServer classes from :core, and provide a simple interface to bring up brokers in integration tests.

Signed-off-by: Arjun Satish <arjunconfluent.io>

Author: Arjun Satish <arjun@confluent.io>
Author: Arjun Satish <wicknicks@users.noreply.github.com>

Reviewers: Randall Hauch <rhauch@gmail.com>, Konstantine Karantasis <konstantine@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5516 from wicknicks/connect-integration-test

tombentley committed:
Fix KAFKA-7789 by increasing the key size for the RSA keys generated for (#6096)

Reviewers: Jun Rao <junrao@gmail.com>

AndrewJSchofield committed:
KAFKA-7461: Add tests for logical types

Added testing of logical types for Kafka Connect in support of KIP-145 features.
Added tests for Boolean, Time, Date and Timestamp, including the valid conversions.

The area of ISO8601 strings is a bit of a mess because the tokenizer is not compatible with
that format, and a subsequent JIRA will be needed to fix that.

A few small fixes as well as creating test cases, but they're clearly just corrections such as
using 0 to mean January (java.util.Calendar uses zero-based month numbers).

Author: Andrew Schofield <andrew_schofield@uk.ibm.com>

Reviewers: Mickael Maison <mimaison@users.noreply.github.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #6077 from AndrewJSchofield/KAFKA-7461-ConverterValuesLogicalTypesTest

stanislavkozlovski committed:
MINOR: Update Trogdor StringExpander regex to handle an epilogue (#6123)

Update the Trogdor StringExpander regex to handle an epilogue.  Previously the regex would use a lazy quantifier at the end, which meant it would not catch anything after the range expression.  Add a unit test.

Reviewed-by: Colin P. McCabe <cmccabe@apache.org>

chia7712 committed:
KAFKA-7391; Introduce close(Duration) to Producer and AdminClient instead of close(long, TimeUnit) (#5667)

See KIP-367: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=89070496.

Reviewers: Viktor Somogyi <viktorsomogyi@gmail.com>, Manikumar Reddy <manikumar.reddy@gmail.com>, Jason Gustafson <jason@confluent.io>

guozhangwang committed:
MINOR: Add log entry for KafkaException in StreamThread#runLoop (#6144)

I've observed several reports of sudden unexpected streamthread shutdown with the log entry like:

State transition from PENDING_SHUTDOWN to DEAD

but there is no related error logs before this line at all. I suspect this is because we intentionally do not log for KafkaException and there's some edge cases where we miss internally and hence caused this. I'm adding the ERROR level log entry here in order to reveal more information in case I saw this again in the future.

Reviewers: Matthias J. Sax <matthias@confluent.io>

guozhangwang committed:
MINOR: Streams upgrade guide section for newly merged KIPs (#6108)

Reviewers: Bill Bejeck <bbejeck@gmail.com>, Matthias J. Sax <matthias@confluent.io>

stanislavkozlovski committed:
KAFKA-7792: Add simple /agent/uptime and /coordinator/uptime health check endpoints (#6130)

Reviewed-by: Colin P. McCabe <cmccabe@apache.org>

Commit:
KAFKA-7824; Require member.id for initial join group request [KIP-394] (#6058)

This patch implements KIP-394 as documented in https://cwiki.apache.org/confluence/display/KAFKA/KIP-394%3A+Require+member.id+for+initial+join+group+request.

Reviewers: Stanislav Kozlovski <stanislav_kozlovski@outlook.com>, Jason Gustafson <jason@confluent.io>

mageshn committed:
MINOR: Start Connect REST server in standalone mode to match distributed mode (KAFKA-7503 follow-up)

Start the Rest server in the standalone mode similar to how it's done for distributed mode.

Author: Magesh Nandakumar <magesh.n.kumar@gmail.com>

Reviewers: Arjun Satish <arjun@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #6148 from mageshn/KAFKA-7826

mumrah committed:
KAFKA-7738; Track leader epochs in client Metadata (#6045)

Track the last seen partition epoch in the Metadata class. When handling metadata updates, check that the partition info being received is for the last seen epoch or a newer one. This prevents stale metadata from being loaded into the client.

Reviewers: Jason Gustafson <jason@confluent.io>

lindong28 committed:
KAFKA-7829; Javadoc should show that AdminClient.alterReplicaLogDirs() is supported in Kafka 1.1.0 or later (#6157)

Reviewers: Jun Rao <junrao@gmail.com>

dhruvilshah3 committed:
MINOR: log when controller begins processing logdir failure event (#6153)

Reviewers: Jun Rao <junrao@gmail.com>

yaojingguo committed:
MINOR:: Fix typos (#6079)

1. Use singular form instead of plural form
2. Add a missing period

ConcurrencyPractitioner committed:
[KAFKA-7024] Rocksdb state directory should be created before opening the DB (#6138)

In RocksDBStore.openDB we call

Files.createDirectories(dir.getParentFile().toPath()); 
return RocksDB.open(options, dir.getAbsolutePath());

We would also add the absolute file path as well to avoid the extra logging.

Reviewers: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

guozhangwang committed:
MINOR: Remove the InvalidTopicException handling in InternalTopicManager (#6167)

Note we can only remove this handling in 2.2 but not in 2.1 since #6124 is only in 2.2.

Reviewers: Bill Bejeck <bill@confluent.io>, Damian Guy <damian@confluent.io>, Matthias J. Sax <matthias@confluent.io>

guozhangwang committed:
KAFKA-7652: Part I; Fix SessionStore's findSession(single-key) (#6134)

Let findSessions(final K key) to call on underlying bytes store directly, using the more restricted range.

Fix the conservative upper range for multi-key range in session schema.

Minor: removed unnecessary private WrappedSessionStoreBytesIterator class as it is only used in unit test.

Minor: removed unnecessary schema#init function by using the direct bytes-to-binary function.

Please read the original PR for more detailed explanation of the root cause of the bug.


Reviewers: Bill Bejeck <bill@confluent.io>, Damian Guy <damian@confluent.io>, John Roesler <john@confluent.io>

jaceklaskowski committed:
MINOR: Log partition info when creating new request batch in controller (#6145)

Due to the missing `$`, the name was being logged instead of the value.

Reviewers: Ismael Juma <ismael@juma.me.uk>

mjsax committed:
KAFKA-6455: Extend CacheFlushListener to forward timestamp (#6147)

Reviewers: Bill Bejeck <bill@confluent.io>, Guozhang Wang <guozhang@confluent.io>

lfrancke committed:
Fix Javadoc of KafkaConsumer (#6155)

The Javadoc is using Properties.put which should never be used because it allows putting non-strings into a Properties object which is designed to only handle strings.
Two other minor fixes so the examples actually work

mjsax committed:
KAFKA-3522: Add internal RecordConverter interface (#6150)

Reviewers: Bill Bejeck <bill@confluent.io>, Guozhang Wang <guozhang@confluent.io>

dongjinleekr committed:
MINOR: Remove unused imports, exceptions, and values (#6117)

1. Remove unthrown exceptions from MemoryRecordsBuilderTest
2. Remove unused imports from ReplicaFetcherThread, ZooKeeperClient, ApiVersionTest, PartitionTest
3. Remove unused value from PartitionTest

wicknicks committed:
MINOR: Handle case where connector status endpoints returns 404 (#6176)

Reviewers: Randall Hauch <randall@confluent.io>, Matthias J. Sax <matthias@confluent.io>

Commit:
MINOR: Rejoin split ssl principal mapping rules (#6099)

* Join ssl principal mapping rules correctly before evaluating.

Java properties splits the configuration array on commas, and that leads to rules containing commas being split before being evaluated. This commit adds a code change to re-join those strings into full rules before evaluating them. The function assumes every rule is either DEFAULT or begins with the prefix RULE:

mjsax committed:
MINOR: increase timeouts for KafkaStreamsTest (#6178)

Reviewer: Arjun Satish <arjun@confluent.io>

dongjinleekr committed:
Fix Documentation for cleanup.policy is out of date (#6181)

stanislavkozlovski committed:
KAFKA-7844: Use regular subproject for generator to fix *All targets (#6182)

The presence of the buildSrc subproject is causing problems when we try
to run installAll, jarAll, and the other "all" targets. It's easier
just to make the generator code a regular subproject and use the
JavaExec gradle task to run the code. This also makes it more
straightforward to run the generator unit tests.

Reviewers: David Arthur <mumrah@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Co-authored-by: Colin P. Mccabe <cmccabe@confluent.io>
Co-authored-by: Stanislav Kozlovski <stanislav_kozlovski@outlook.com>

hachikuji committed:
MINOR: Cleanup handling of mixed transactional/idempotent records (#6172)

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>, Ismael Juma <ismael@juma.me.uk>, Colin Patrick McCabe <colin@cmccabe.xyz>

mjsax commented:
Just a side remark. If this should be part of `2.2` release, we must merge it by feature freeze deadline that is Jan 31.

C0urante committed:
KAFKA-5117: Stop resolving externalized configs in Connect REST API

[KIP-297](https://cwiki.apache.org/confluence/display/KAFKA/KIP-297%3A+Externalizing+Secrets+for+Connect+Configurations#KIP-297:ExternalizingSecretsforConnectConfigurations-PublicInterfaces) introduced the `ConfigProvider` mechanism, which was primarily intended for externalizing secrets provided in connector configurations. However, when querying the Connect REST API for the configuration of a connector or its tasks, those secrets are still exposed. The changes here prevent the Connect REST API from ever exposing resolved configurations in order to address that. rhauch has given a more thorough writeup of the thinking behind this in [KAFKA-5117](https://issues.apache.org/jira/browse/KAFKA-5117)

Tested and verified manually. If these changes are approved unit tests can be added to prevent a regression.

Author: Chris Egerton <chrise@confluent.io>

Reviewers: Robert Yokota <rayokota@gmail.com>, Randall Hauch <rhauch@gmail.com, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #6129 from C0urante/hide-provided-connect-configs

likan999 committed:
MINOR: ducker-ak: add down -f, avoid using a terminal in ducker test

When using ./ducker-ak test on Jenkins, the script complains that there is no TTY.  To fix this, we should skip passing -t to docker exec.  We do not need a pseudo-TTY to run the tests.  Similarly, we should skip passing -i, since we do not need to keep stdin open.

The down command should have a force option, specified as -f or --force.

Reviewed-by: Colin P. McCabe <cmccabe@apache.org>

bbejeck committed:
KAFKA-4850: Enable bloomfilters (#6012)

This PR enables BloomFilters for RocksDB to speed up point lookups.
The request for this has been around for some time - https://issues.apache.org/jira/browse/KAFKA-4850

For testing, I've done the following

Ran the standard streams suite of unit and integration tests
Kicked off the simple benchmark test with bloom filters enabled
Kicked off the simple benchmark test with bloom filters not enabled
Kicked off streams system tests

Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>, John Roesler <john@confluent.io>

mjsax committed:
MINOR: fix race condition in KafkaStreamsTest (#6185)

Reviewers: John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

cmccabe committed:
KAFKA-7793: Improve the Trogdor command line. (#6133)

* Allow the Trogdor agent to be started in "exec mode", where it simply
runs a single task and exits after it is complete.

* For AgentClient and CoordinatorClient, allow the user to pass the path
to a file containing JSON, instead of specifying the JSON object in the
command-line text itself.  This means that we can get rid of the bash
scripts whose only function was to load task specs into a bash string
and run a Trogdor command.

* Print dates and times in a human-readable way, rather than as numbers
of milliseconds.

* When listing tasks or workers, output human-readable tables of
information.

* Allow the user to filter on task ID name, task ID pattern, or task
state.

* Support a --json flag to provide raw JSON output if desired.

Reviewed-by: David Arthur <mumrah@gmail.com>, Stanislav Kozlovski <stanislav_kozlovski@outlook.com>

cmccabe committed:
MINOR: update copyright year in the NOTICE file. (#6196)

mingaliu committed:
KAFKA-7692; Fix ProducerStateManager SequenceNumber overflow (#5990)

This patch fixes a few overflow issues with wrapping sequence numbers in the broker's producer state tracking. 

Reviewers: Jason Gustafson <jason@confluent.io>

mingaliu committed:
KAFKA-7693; Fix SequenceNumber overflow in producer (#5989)

The problem is that the sequence number is an Int and should wrap around when it reaches the Int.MaxValue. The bug here is it doesn't wrap around and become negative and raises an error.

Reviewers: Jason Gustafson <jason@confluent.io>

Commit:
MINOR: upgrade to jdk8 8u202

Upgrade from 171 to 202. Unpack and install directly from a cached tgz rather than going via the installer deb from webupd8. The installer is still on 8u919 while we want 202.

Testing via kafka branch builder job

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2305/

Author: Jarek Rudzinski <jarek@confluent.io>
Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Alex Diachenko <sansanichfb@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #6165 from jarekr/trunk-jdk8-from-tgz

dongjinleekr committed:
MINOR: Improve IntegrationTestUtils documentation (#5664)

* 1. Add Javadoc to undocumented methods. 2. Add documentation on  parameters in  methods. 3. Fix typo and method ordering.

* Remove IntegrationTestUtils#produceKeyValuesSynchronouslyWithTimestamp(String, Collection<KeyValue<K, V>>, Properties, Headers, Long)

kkonstantine committed:
MINOR: Upgrade ducktape to 0.7.5 (#6197)

Reviewed-by: Colin P. McCabe <cmccabe@apache.org>

vvcephei committed:
MINOR: clarify why suppress can sometimes drop tombstones (#6195)

Reviewers: Jonathan Gordon, Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

tombentley committed:
KAFKA-5692: Change PreferredReplicaLeaderElectionCommand to use Admin… (#3848)

See also KIP-183.

This implements the following algorithm:

AdminClient sends ElectPreferredLeadersRequest.
KafakApis receives ElectPreferredLeadersRequest and delegates to
ReplicaManager.electPreferredLeaders()
ReplicaManager delegates to KafkaController.electPreferredLeaders()
KafkaController adds a PreferredReplicaLeaderElection to the EventManager,
ReplicaManager.electPreferredLeaders()'s callback uses the
delayedElectPreferredReplicasPurgatory to wait for the results of the
election to appear in the metadata cache. If there are no results
because of errors, or because the preferred leaders are already leading
the partitions then a response is returned immediately.
In the EventManager work thread the preferred leader is elected as follows:

The EventManager runs PreferredReplicaLeaderElection.process()
process() calls KafkaController.onPreferredReplicaElectionWithResults()
KafkaController.onPreferredReplicaElectionWithResults()
calls the PartitionStateMachine.handleStateChangesWithResults() to
perform the election (asynchronously the PSM will send LeaderAndIsrRequest
to the new and old leaders and UpdateMetadataRequest to all brokers)
then invokes the callback.

Reviewers: Colin P. McCabe <cmccabe@apache.org>, Jun Rao <junrao@gmail.com>

dhruvilshah3 committed:
KAFKA-7838: Log leader and follower end offsets when shrinking ISR (#6168)

Reviewers: Jun Rao <junrao@gmail.com>

gardnervickers committed:
MINOR: In the MetadataResponse schema, ignorable should be a boolean

dhruvilshah3 committed:
KAFKA-7837: Ensure offline partitions are picked up as soon as possible when shrinking ISR (#6202)

Check if a partition is offline while iterating all partitions.

Reviewers: Jun Rao <junrao@gmail.com>

Bert-R committed:
Added quotes around the class path (#4469)

The script otherwise fails when the classpath contains paths with spaces.

mjsax committed:
KAFKA-3522: Generalize Segments (#6170)

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

ijuma committed:
MINOR: Update Gradle to 5.1.1 (#6160)

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

cadonna committed:
KAFKA-4217: Add KStream.flatTransform (#5273)

Reviewers: Matthias J. Sax <matthias@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

mjsax committed:
MINOR: Update usage of deprecated API (#6146)

Reviewers: Guozhang Wang <guozhang@confluent.io>, Ismael Juma <ismael@confluent.io>, Jorge Quilcate Otoya <quilcate.jorge@gmail.com>, John Roesler <john@confluent.io>, Bill Bejeck <bill@confluent.io>

mjsax committed:
KAFKA-3522: Replace RecordConverter with TimestampedBytesStore (#6204)

Reviewers: John Roesler <john@confluent.io>, Bill Bejeck <bill@confluent.io>, Guozhang Wang <guozhang@confluent.io>

mjsax committed:
KAFKA-3522: Add RocksDBTimestampedStore (#6149)

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

guozhangwang committed:
KAFKA-7652: Part II; Add single-point query for SessionStore and use for flushing / getter (#6161)

#2972 tried to fix a bug about flushing operation, but it was not complete, since findSessions(key, earliestEnd, latestStart) does not guarantee to only return a single entry since its semantics are to return any sessions whose end > earliestEnd and whose start < latestStart.

I've tried various ways to fix it completely and I ended up having to add a single-point query to the public ReadOnlySessionStore API for the exact needed semantics. It is used for flushing to read the old values (otherwise the wrong old values will be sent downstreams, hence it is a correctness issue) and also for getting the value for value-getters (it is for perf only).

Commit:
KAFKA-7859: Use automatic RPC generation in LeaveGroups (#6188)

Reviewed-by: Colin P. McCabe <cmccabe@apache.org>

chia7712 committed:
MINOR: fix checkstyle suppressions for generated RPC code to work on Windows

Reviewed-by: Colin P. McCabe <cmccabe@apache.org>

rajinisivaram committed:
KAFKA-7719: Improve fairness in SocketServer processors (KIP-402) (#6022)

Limit the number of new connections processed in each iteration of each
Processor. Block Acceptor if the connection queue is full on all Processors.
Added a metric to track accept blocked time percent. See KIP-402 for details.

Reviewers: Ismael Juma <ismael@juma.me.uk>

rhauch committed:
KAFKA-7873; Always seek to beginning in KafkaBasedLog (#6203)

Explicitly seek KafkaBasedLog’s consumer to the beginning of the topic partitions, rather than potentially use committed offsets (which would be unexpected) if group.id is set or rely upon `auto.offset.reset=earliest` if the group.id is null.

This should not change existing behavior but should remove some potential issues introduced with KIP-287 if `group.id` is not set in the consumer configurations. Note that even if `group.id` is set, we still always want to consume from the beginning.

Reviewers: Jason Gustafson <jason@confluent.io>

cmccabe committed:
MINOR: Fix some field definitions for ListOffsetReponse (#6214)

Reviewers: Jun Rao <junrao@gmail.com>

viktorsomogyi committed:
KAFKA-7433; Introduce broker options in TopicCommand to use AdminClient (KIP-377)

The PR adds --bootstrap-server and --admin.config options to TopicCommand and implements an alternative, AdminClient based way of topic management.

As testing I've duplicated the existing tests and made them working with the AdminClient options.

Author: Viktor Somogyi-Vass <viktorsomogyi@gmail.com>

Reviewers: Andras Katona <41361962+akatona84@users.noreply.github.com>, Sandor Murakozi <smurakozi@gmail.com>, Manikumar Reddy <manikumar.reddy@gmail.com>, Jason Gustafson <jason@confluent.io>

Closes #5683 from viktorsomogyi/topiccommand-adminclient

stanislavkozlovski committed:
KAFKA-7641; Introduce "group.max.size" config to limit group sizes (#6163)

This patch introduces a new config - "group.max.size", which caps the maximum size any group can reach. It has a default value of Int.MAX_VALUE. Once a group is of the maximum size, subsequent JoinGroup requests receive a MAX_SIZE_REACHED error.

In the case where the config is changed and a Coordinator broker with the new config loads an old group that is over the threshold, members are kicked out of the group and a rebalance is forced.

Reviewers: Vahid Hashemian <vahid.hashemian@gmail.com>, Boyang Chen <bchen11@outlook.com>, Gwen Shapira <cshapi@gmail.com>, Jason Gustafson <jason@confluent.io>

guozhangwang committed:
KAFKA-7798: Expose embedded clientIds (#6107)

Reviewers: Damian Guy <damian@confluent.io>, John Roesler <john@confluent.io>, Boyang Chen <bchen11@outlook.com>, Matthias J. Sax <matthias@confluent.io>, Bill Bejeck <bill@confluent.io>

mumrah committed:
KAFKA-7766: Fail fast PR builds (#6059)

Split the Gradle invocation in the jenkins.sh script into two commands so
we can fail fast for validation checks such as compile errors and checkstyle
errors.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Ismael Juma <ismael@juma.me.uk>

stanislavkozlovski committed:
MINOR: Reduce replica.fetch.backoff.ms in ReassignPartitionsClusterTest (#5887)

The default backoff of 1000ms when there are no partitions to fetch can cause `shouldExecuteThrottledReassignment` to fail due to it taking too long. So we reduce
it to 100ms.

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>, Ismael Juma <ismael@juma.me.uk

lambdaliu committed:
fix typo (#5150)

cmccabe committed:
KAFKA-7832: Use automatic RPC generation in CreateTopics (#5972)

Reviewers: Jun Rao <junrao@gmail.com>, Tom Bentley <tbentley@redhat.com>, Boyang Chen <bchen11@outlook.com>

viktorsomogyi committed:
KAFKA-7804: Update docs for topic-command related KIP-377

This PR adds a upgrade notes and changes examples to use the bootstrap-server.

Author: Viktor Somogyi-Vass <viktorsomogyi@gmail.com>

Reviewers: Srinivas <srinivas96alluri@gmail.com>, Manikumar Reddy <manikumar.reddy@gmail.com>

Closes #6118 from viktorsomogyi/topiccommand-adminclient-doc

cmccabe committed:
MINOR: release.py: fix some compatibility problems.

Rather than using sed, use built-in Python regular expressions to strip
the SNAPSHOT expression from the pom.xml files.  Sed has different flags
on different platforms, such as Linux.  Using Python directly here is
more compatible, as well as being more efficient, and not requiring an
rm command afterwards.

When running release_notes.py, use the current Python interpreter.
This is needed to prevent attempting to run release_notes.py with
Python 3 on some systems.  release_notes.py will not (yet) work with
Python 3.

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Magnus Edenhill <magnus@edenhill.se>, David Arthur <mumrah@gmail.com>, Manikumar Reddy <manikumar.reddy@gmail.com>

Closes #6198 from cmccabe/release_py

kkonstantine committed:
KAFKA-7834: Extend collected logs in system test services to include heap dumps

* Enable heap dumps on OOM with -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=<file.bin> in the major services in system tests
* Collect the heap dump from the predefined location as part of the result logs for each service
* Change Connect service to delete the whole root directory instead of individual expected files
* Tested by running the full suite of system tests

Author: Konstantine Karantasis <konstantine@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #6158 from kkonstantine/KAFKA-7834

markcho committed:
KAFKA-7890: Invalidate ClusterConnectionState cache for a broker if the hostname of the broker changes. (#6215)

cmccabe committed:
KAFKA-7828: Add ExternalCommandWorker to Trogdor (#6219)

Allow the Trogdor agent to execute external commands. The agent communicates with the external commands via stdin, stdout, and stderr.

Based on a patch by Xi Yang <xi@confluent.io>

Reviewers: David Arthur <mumrah@gmail.com>

hachikuji committed:
KAFKA-7540; Retry coordinator lookup to fix transient failure in ConsumerBounceTest (#6235)

Add logic in ConsumerBounceTest to check the error code in FindCoordinator responses and retry if needed. This should help with transient failures or at least get us closer to the actual problem.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

soenkeliebau committed:
MINOR: Add sentence to command line documentation of producer performance tool on how to disable throttling. (#6205)

rondagostino committed:
KAFKA-7902: Replace original loginContext if SASL/OAUTHBEARER refresh login fails (#6233)

Replaces original loginContext if login fails in the refresh thread to ensure that the refresh thread is left in a clean state when there are exceptions while connecting to an OAuth server. Also makes client callback handler more robust by using the token with the longest remaining time for expiry instead of throwing an exception if multiple tokens are found.

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>

hachikuji committed:
KAFKA-7897; Disable leader epoch cache when older message formats are used (#6232)

When an older message format is in use, we should disable the leader epoch cache so that we resort to truncation by high watermark. Previously we updated the cache for all versions when a broker became leader for a partition. This can cause large and unnecessary truncations after leader changes because we relied on the presence of _any_ cached epoch in order to tell whether to use the improved truncation logic possible with the OffsetsForLeaderEpoch API.

Reviewers: Stanislav Kozlovski <stanislav_kozlovski@outlook.com>, Viktor Somogyi-Vass <viktorsomogyi@gmail.com>, Jun Rao <junrao@gmail.com>

mjsax committed:
MINOR: improve stabilty of ProcessorStateManagerTest (#6240)

This PR addressed the following test failure:

```
java.lang.AssertionError:
Expected: a string starting with "process-state-manager-test Failed to write offset checkpoint file to ["
but: was "[AdminClient clientId=adminclient-874] Connection to node -1 (localhost/127.0.0.1:8080) could not be established. Broker may not be available."
```

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

radai-rosenblatt committed:
MINOR: Include partition when logging fetch errors (#6206)

This helps narrow down the specific broker they came from when debugging
ACL propagation issues.

Reviewers: Vahid Hashemian <vahid.hashemian@gmail.com>, Ismael Juma <ismael@juma.me.uk>

ouertani committed:
format file

ouertani committed:
format file

ouertani committed:
rebase trunk

ouertani commented:
@guozhangwang @ryannedolan updates done

mjsax commented:
@ouertani This PR cannot be review atm -- it needs to be rebased to `trunk` first. Thanks.

mjsax commented:
\cc @vvcephei for review (after PR is rebased).

casey-green commented:
I tried using this today, so +1000 on getting this change in 😄

guozhangwang commented:
@ouertani are you still working on this PR? If yes could you rebase it and then ping us for reviews?

casey-green commented:
@guozhangwang @vvcephei I created a different PR for this since this one seems stuck https://github.com/apache/kafka/pull/6314

Unit tests are included in that one!

mjsax commented:
What is the status of this PR?

casey-green commented:
@mjsax this can be closed, I implemented this functionality in #6314

kkonstantine commented:
Closing in favor of https://github.com/apache/kafka/pull/6314
